### PR TITLE
Initial journald log provider attempt

### PR DESCRIPTION
**Pull request by LucasRoesler**
## Description

- journald log provider using exec to journalctl
```
journalctl -t <namespace>:<name>  --output=json --since=<timestamp> <--follow> --output-fields=SYSLOG_IDENTIFIER,MESSAGE,_PID,_SOURCE_REALTIME_TIMESTAMP
```

## Motivation and Context

- [x] I have raised an issue to propose this change **this is required**

## How Has This Been Tested?

I have tested it manually using a multipass instance

```sh
ubuntu@primary:~$ sudo cp /home/ubuntu/Home/Code/go/src/github.com/openfaas/faasd/bin/faasd /usr/local/bin/faasd
(reverse-i-search)`pk': sudo ^Cill faasd -9 -e
ubuntu@primary:~$ ^C
ubuntu@primary:~$ ^C
ubuntu@primary:~$ ^C
ubuntu@primary:~$ faas-cli list
Function                      	Invocations    	Replicas
nodeinfo                      	0              	1
ping                          	0              	0
ubuntu@primary:~$ faas-cli logs nodeinfo
WARNING! Communication is not secure, please consider using HTTPS. Letsencrypt.org offers free SSL/TLS certificates.
2020-03-04T21:04:32+01:00 Forking - node [index.js]
2020-03-04T21:04:32+01:00 2020/03/04 20:04:32 Started logging stderr from function.
2020-03-04T21:04:32+01:00 2020/03/04 20:04:32 Started logging stdout from function.
2020-03-04T21:04:32+01:00 2020/03/04 20:04:32 OperationalMode: http
2020-03-04T21:04:32+01:00 2020/03/04 20:04:32 Timeouts: read: 15s, write: 15s hard: 10s.
2020-03-04T21:04:32+01:00 2020/03/04 20:04:32 Listening on port: 8080
2020-03-04T21:04:32+01:00 2020/03/04 20:04:32 Writing lock-file to: /tmp/.lock
2020-03-04T21:04:32+01:00 2020/03/04 20:04:32 Metrics listening on port: 8081
2020-03-04T21:04:32+01:00 2020/03/04 20:04:32 stdout: OpenFaaS Node.js listening on port: 3000
2020-03-04T21:04:33+01:00 2020/03/04 20:04:33 POST / - 200 OK - ContentLength: 84
2020-03-04T21:04:34+01:00 2020/03/04 20:04:34 POST / - 200 OK - ContentLength: 84
2020-03-04T21:14:45+01:00 2020/03/04 20:14:45 POST / - 200 OK - ContentLength: 84
2020-03-04T21:14:46+01:00 2020/03/04 20:14:46 POST / - 200 OK - ContentLength: 84
2020-03-04T21:14:47+01:00 2020/03/04 20:14:47 POST / - 200 OK - ContentLength: 84
2020-03-04T21:19:24+01:00 Forking - node [index.js]
2020-03-04T21:19:24+01:00 2020/03/04 20:19:24 Started logging stderr from function.
2020-03-04T21:19:24+01:00 2020/03/04 20:19:24 Started logging stdout from function.
2020-03-04T21:19:24+01:00 2020/03/04 20:19:24 OperationalMode: http
2020-03-04T21:19:24+01:00 2020/03/04 20:19:24 Timeouts: read: 15s, write: 15s hard: 10s.
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

Commits:

- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] My commit message has a body and describe how this was tested and why it is required.
- [x] I have signed-off my commits with `git commit -s` for the Developer Certificate of Origin (DCO)

Code:

- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.

Docs:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,6 +8,7 @@ addons:
   apt:
     packages:
       - runc
+      - libsystemd-dev
```
Comment (alexellis) [inline .travis.yml]: Why is this required? Is it part of the original CGO version of the PR?

File: Makefile [hunk 2]
```diff
@@ -9,7 +9,7 @@ ARCH := amd64
 all: local
 
 local:
-	CGO_ENABLED=0 GOOS=linux go build -o bin/faasd
+	CGO_ENABLED=0 GOOS=linux go build -ldflags $(LDFLAGS) -v -o bin/faasd
```
Comment (alexellis) [inline Makefile]: Is this required for make local? It slows the build down considerably.

File: pkg/logs/requestor_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+	expectedArgs := fmt.Sprintf(
+		"--utc --no-pager --output=json --output-fields=SYSLOG_IDENTIFIER,MESSAGE,_PID,_SOURCE_REALTIME_TIMESTAMP --identifier=spacetwo:loggyfunc --since=%s --follow --lines=5",
+		now.UTC().Format("2006-01-02 15:04:05"),
+	)
+
+	cmd := buildCmd(ctx, req).String()
+
+	if !strings.Contains(cmd, "journalctl") {
+		t.Fatalf("expected journalctl cmd, ngot cmd %s", cmd)
```
Comment (alexellis) [inline pkg/logs/requestor_test.go]: `ngot` or `\ngot`?

File: Makefile (hunk 2 above)
Comment (LucasRoesler) [reply Makefile]: the -v is not but i found it useful to see what was actually happening  and to also make the cmd easier to debug because the version would actually update as a commit

File: .travis.yml (hunk 1 above)
Comment (LucasRoesler) [reply .travis.yml]: you are right, i will remove it

**Comment by LucasRoesler** [comment]
There are some really basic unit tests, which could be expanded to cover more branches. But I have tested it manually. The big thing that is missing is that I need to update the commit with test instructions per the PR checklist.

File: Makefile (hunk 2 above)
Comment (alexellis) [reply Makefile]: -v is fine, but the ldflags I meant

**Comment by alexellis** [comment]
Can you please add to the e2e tests? The absolute basic use-case is fine i.e. invoke followed by grep.

**Comment by alexellis** [comment]
You may be surprised how simple these are, but they have saved a few bugs going in already

https://github.com/openfaas/faasd/blob/master/Makefile#L38

**Comment by LucasRoesler** [comment]
I just updated the commit with the additional e2e test ... now we wait :)

**Comment by alexellis** [comment]
Screenshot 2020-03-06 at 12 45 13

You need an input to the command ☝️

**Comment by LucasRoesler** [comment]
🤦‍♂

**Comment by alexellis** [comment]
Seeing an X again

**Comment by LucasRoesler** [comment]
Now i don't know what is happening, it was working before I rebased on the latest master :(

Now there are no error logs in the output it is just stuck

```sh
Mar 07 08:44:02 primary faasd[3867]: 2020/03/07 09:44:02 Resolve: "nodeinfo"
Mar 07 08:44:02 primary faasd[3867]: 2020/03/07 09:44:02 nodeinfo took 0.122992 seconds
Mar 07 08:44:03 primary faasd[3867]: 2020/03/07 09:44:03 Resolve: "nodeinfo"
Mar 07 08:44:03 primary faasd[3867]: 2020/03/07 09:44:03 nodeinfo took 0.004651 seconds
Mar 07 08:44:09 primary faasd[3867]: 2020/03/07 09:44:09 starting journal stream using  /bin/journalctl --utc --no-pager --output=json --output-fields=SYSLOG_IDENTIFIER,MESSAGE,_PID,_SOURCE_REALTIME_TIMESTAMP --identifier=openfaas-fn:nodeinfo --since=2020-03-07 08:39:09 --follow
Mar 07 08:45:09 primary faasd[3867]: 2020/03/07 09:45:09 LogHandler: client stopped listening
Mar 07 08:45:09 primary faasd[3867]: 2020/03/07 09:45:09 closing journal stream
Mar 07 08:45:09 primary faasd[3867]: 2020/03/07 09:45:09 stderr closed
Mar 07 08:45:09 primary faasd[3867]: 2020/03/07 09:45:09 wait result signal: killed
```

I will keep debugging

**Comment by LucasRoesler** [comment]
wtf, all i did was add a log line and it is starting to work 
```diff
--- pkg/logs/requestor.go
+++ pkg/logs/requestor.go
@@ -138,6 +138,7 @@ func streamLogs(ctx context.Context, cmd *exec.Cmd, out io.ReadCloser, msgs chan
                        return
                }

+               log.Printf("sending msg %+v\nfor entry %+v\n", msg, entry)
                msgs <- msg
        }
 }
```

```sh
ubuntu@primary:~$ faas-cli list
Function                      	Invocations    	Replicas
figlet                        	0              	0
nodeinfo                      	0              	0
ping                          	0              	0
ubuntu@primary:~$ faas-cli invoke nodeinfo <<< adf
Server returned unexpected status code: 500 -
ubuntu@primary:~$ faas-cli invoke nodeinfo <<< adf
Hostname: localhost

Arch: x64
CPUs: 1
Total mem: 986MB
Platform: linux
Uptime: 74
ubuntu@primary:~$ faas-cli invoke nodeinfo <<< adf
Hostname: localhost

Arch: x64
CPUs: 1
Total mem: 986MB
Platform: linux
Uptime: 75
ubuntu@primary:~$ faas-cli logs nodeinfo
WARNING! Communication is not secure, please consider using HTTPS. Letsencrypt.org offers free SSL/TLS certificates.
2020-03-07T08:43:59Z Forking - node [index.js]
2020-03-07T08:43:59Z 2020/03/07 08:43:59 Started logging stderr from function.
2020-03-07T08:43:59Z 2020/03/07 08:43:59 Started logging stdout from function.
2020-03-07T08:43:59Z 2020/03/07 08:43:59 OperationalMode: http
2020-03-07T08:43:59Z 2020/03/07 08:43:59 Timeouts: read: 15s, write: 15s hard: 10s.
2020-03-07T08:43:59Z 2020/03/07 08:43:59 Listening on port: 8080
2020-03-07T08:43:59Z 2020/03/07 08:43:59 Writing lock-file to: /tmp/.lock
2020-03-07T08:43:59Z 2020/03/07 08:43:59 Metrics listening on port: 8081
2020-03-07T08:44:00Z 2020/03/07 08:44:00 stdout: OpenFaaS Node.js listening on port: 3000
2020-03-07T08:44:02Z 2020/03/07 08:44:02 POST / - 200 OK - ContentLength: 86
2020-03-07T08:44:03Z 2020/03/07 08:44:03 POST / - 200 OK - ContentLength: 86
2020-03-07T08:57:08Z Forking - node [index.js]
2020-03-07T08:57:08Z 2020/03/07 08:57:08 Started logging stderr from function.
2020-03-07T08:57:08Z 2020/03/07 08:57:08 Started logging stdout from function.
2020-03-07T08:57:08Z 2020/03/07 08:57:08 OperationalMode: http
2020-03-07T08:57:08Z 2020/03/07 08:57:08 Timeouts: read: 15s, write: 15s hard: 10s.
2020-03-07T08:57:08Z 2020/03/07 08:57:08 Listening on port: 8080
2020-03-07T08:57:08Z 2020/03/07 08:57:08 Writing lock-file to: /tmp/.lock
2020-03-07T08:57:08Z 2020/03/07 08:57:08 Metrics listening on port: 8081
```

**Comment by alexellis** [comment]
logserr

**Comment by LucasRoesler** [comment]
is it possible there are a lot of variations of journalctl?

```sh
ubuntu@primary:~$ journalctl --version
systemd 237
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid
```

**Comment by alexellis** [comment]
```
journalctl [OPTIONS...] [MATCHES...]

Query the journal.

Options:
     --system              Show the system journal
     --user                Show the user journal for the current user
  -M --machine=CONTAINER   Operate on local container
  -S --since=DATE          Show entries not older than the specified date
  -U --until=DATE          Show entries not newer than the specified date
  -c --cursor=CURSOR       Show entries starting at the specified cursor
     --after-cursor=CURSOR Show entries after the specified cursor
     --show-cursor         Print the cursor after all the entries
  -b --boot[=ID]           Show current boot or the specified boot
     --list-boots          Show terse information about recorded boots
  -k --dmesg               Show kernel message log from the current boot
  -u --unit=UNIT           Show logs from the specified unit
     --user-unit=UNIT      Show logs from the specified user unit
  -t --identifier=STRING   Show entries with the specified syslog identifier
  -p --priority=RANGE      Show entries with the specified priority
  -e --pager-end           Immediately jump to the end in the pager
  -f --follow              Follow the journal
  -n --lines[=INTEGER]     Number of journal entries to show
     --no-tail             Show all lines, even in follow mode
  -r --reverse             Show the newest entries first
  -o --output=STRING       Change journal output mode (short, short-iso,
                                   short-precise, short-monotonic, verbose,
                                   export, json, json-pretty, json-sse, cat)
     --utc                 Express time in Coordinated Universal Time (UTC)
  -x --catalog             Add message explanations where available
     --no-full             Ellipsize fields
  -a --all                 Show all fields, including long and unprintable
  -q --quiet               Do not show info messages and privilege warning
     --no-pager            Do not pipe output into a pager
  -m --merge               Show entries from all available journals
  -D --directory=PATH      Show journal files from directory
     --file=PATH           Show journal file
     --root=ROOT           Operate on catalog files below a root directory
     --interval=TIME       Time interval for changing the FSS sealing key
     --verify-key=KEY      Specify FSS verification key
     --force               Override of the FSS key pair with --setup-keys

Commands:
  -h --help                Show this help text
     --version             Show package version
```

**Comment by LucasRoesler** [comment]
can you also share your `--version` output?

**Comment by alexellis** [comment]
```
systemd 229
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN
```

**Comment by alexellis** [comment]
And on my RPi:

```
 journalctl --version
systemd 241 (241)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid
```

**Comment by LucasRoesler** [comment]
Adding `time.Sleep(time.Millisecond)` seems to fix my issue with figlet. I can remove the `output-fields` flag to extend compatibility

File: pkg/logs/requestor.go [hunk 4]
```diff
@@ -103,6 +102,10 @@ func buildCmd(ctx context.Context, req logs.Request) *exec.Cmd {
 // the loop is based on the Decoder example in the docs
 // https://golang.org/pkg/encoding/json/#Decoder.Decode
 func streamLogs(ctx context.Context, cmd *exec.Cmd, out io.ReadCloser, msgs chan logs.Message) {
+	// without this sleep the channel seems to get stuck. This results in either no log messages
+	// being read by the Handler _or_ the messages are read but only flushed when the request
+	// timesout
+	time.Sleep(time.Millisecond)
```
Comment (alexellis) [inline pkg/logs/requestor.go]: Is there a way to avoid this? My gut is telling me "I fixed an issue by adding a sleep" is always the wrong solution - wrong because timing bugs eventually break and in a really intermittent way.

Also OK to merge this and get the feature out to people, if we can spend a little time looking into a solution after merging?

Comment (LucasRoesler) [reply pkg/logs/requestor.go]: i will keep looking for what the underlying cause is, but i don't know what specific part is broken. If you have any suggestions, i am more than happy to remove this
